homeassistant/components/mqtt/mixins.py

```diff
@@ -1218,8 +1225,6 @@ async def async_added_to_hass(self) -> None:
         self._prepare_subscribe_topics()
         await self._subscribe_topics()
         await self.mqtt_async_added_to_hass()
-        if self._discovery_data is not None:
-            send_discovery_done(self.hass, self._discovery_data)
```

Comment:
The bug here is that we were sending the discovery done before the state was written in `add_to_platform_finish`